typechecker+codegen: Emit new bodies for bindings with different types

cg-jl opened the pull request:
Fixes cases where the body is not duplicated when it should,
specially with instantiations that are inferred inside the body for
functions that support both instantiations and overloading. C++ can deal
with overloading, but we add the template parameters ourselves, which
makes the typing wrong for at least one pattern.

Introduces `BindingKey`, `MatchBuilder` and `BindingKeyBuilder`, which
together help keep match body code duplication at a low rate while being
as least intrusive as possible to the logic for each match typecheck:
`typecheck_match` has three big `for` loops with bodies that walk the
match cases in very different ways, and the deduplication system is
orthogonal to all of them. Most of it is implemented between
`MatchBuilder` and `BindingKeyBuilder`.

These builder states avoid creating scopes and duplicate typechecker
work, resulting in ~14% performance increase and 20MB less peak memory
usage when typechecking selfhost:

![image](https://github.com/user-attachments/assets/d2a5cc47-244a-4995-9ee3-43765ab6e879)

`poop` is https://github.com/andrewrk/poop .
`stage1_main` is the result of compiling stage1 from d65f014cc54b986f629fe676d914af01d442b9f7.

cg-jl commented:
I'm asking for more help to review the document. I'll be redoing it since the last three subsections are actually the design doc, which is hard to read, and the rest could go later as a way to describe the constraints of the problem and why those constraints.

alimpfard commented on selfhost/typechecker.jakt:
```diff
@@ -7492,9 +7494,9 @@ struct Typechecker {
     }
 
-    fn typecheck_var_decl(mut this, var: ParsedVarDecl, init: ParsedExpression, scope_id: ScopeId, safety_mode: SafetyMode, span: Span) throws -> CheckedStatement {
+    fn typecheck_var_decl_without_adding(mut this, var: ParsedVarDecl, init: ParsedExpression, scope_id: ScopeId, safety_mode: SafetyMode, span: Span) throws -> (CheckedVariable, CheckedExpression)? {
```
nit: adding what?

cg-jl replied:
Without... registering the variable in a scope. I'll rename the other version to `typecheck_and_register_var_decl`, or inline it if it's only used in one place.

alimpfard replied:
Sure, any names works so long as it's less ambiguous

alimpfard commented on selfhost/typechecker.jakt:
```diff
@@ -12681,3 +12661,159 @@ fn dump_scope(anon scope_id: ScopeId, anon program: &CheckedProgram, indent: i64
+        }
+    }
+
+    // NOTE: This could require IntoIterator, but Range (0..cases.size() in Empty case above) is Iterable, not IntoIterator.
```
🤔 aside: maybe we should make `Iterable<T>` implement `IntoIterator<T>`?

cg-jl replied:
Just saw #1603. If you end up finding a solution you're comfortable with I'll wait for your fix and then fix this immediately, otherwise I can leave it as a FIXME.

alimpfard replied:
Nah, that's a big mess I don't want to get into right now

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+Notice that the first body of the match accepts both types of `a` and `b`
+(`i32` and `u64`):
+
+- `a + b` typechecks as long as `a` implements `Add` with `b` as its RHS.
```
```suggestion
- `a + b` typechecks as long as the type of `a` implements `Add` with `b` as its RHS.
```

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+respective body.
+
+When generating code for a `match`, codegen sees a `CheckedExpression::Match`
+containing a list of `CheckedMatchCase`s. Each `CheckedMatchCase` exactly one
```
```suggestion
containing a list of `CheckedMatchCase`s. Each `CheckedMatchCase` contains exactly one
```

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+large, and the bodies were duplicated many times.
+
+To ensure that similar functions to `typecheck_binary_operation` do not make
+the C++ compiler sit through 18x the amount of Jakt code, we'll add a second
```
```suggestion
the C++ compiler sift through 18x the amount of Jakt code, we'll add a second
```
?

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+
+The problem with this approach is that now the typechecker portion of this
+solution does not respect Constraint 1, since the last checked body is not
+enough to represent the case if bindings from the same parsed pattern group can
```
```suggestion
enough to represent the case if bindings from the same parsed pattern group
```

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+enough to represent the case if bindings from the same parsed pattern group can
+have different binding types.
+
+This makes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail
+codegen, since it calls `maybe_templated<JaktArray<i32>>` on all cases, when
```
```suggestion
This causes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail,
since it codegens a call to `maybe_templated<JaktArray<i32>>` on all cases, when
```

alimpfard commented on documentation/pattern-match-avoid-duplication.md:
```diff
@@ -0,0 +1,266 @@
+To enforce that the set of cases checked by `BindingKeyBuilder` are the correct
+set, and that context is not lost when reading through the long loop bodies of
+`.typecheck_match`, all the state regarding match typecheck is in
+`MatchBuilder`. This struct has two functionalities: - It will make us
```
where second functionality

cg-jl replied:
oh, I incorrectly removed the bullet point for the second paragraph.

alimpfard reviewed (commented):
I'm not sure what you mean by it being hard to read, I think it's pretty comprehensive and easy to read (ignoring some grammatical mistakes, will point out a couple I noticed).

And yeah, it'd be nicer to move the "history" after the current design doc I think - assuming we're treating this as mainly a design doc and not a history post :)

cg-jl commented:
I'll do the rest of things that require access to the repo this afternoon. Thanks for the comprehensive review!

cg-jl commented:
I've adjusted the comments in typechecker since I documented the functions before I introduced the BindingKeys. I also missed an annotation I used to remind myself to make the design doc
